### PR TITLE
BUG: Simplified fix, Overflow in tan and tanh for large complex values.

### LINKED ISSUE #2321: Error in tanh for large complex argument (Trac #1726)
_Original ticket http://projects.scipy.org/numpy/ticket/1726 on 2011-01-28 by trac user mbakker, assigned to unknown._

tanh computes nans for large complex argument:

In [85]: tanh(1000+0j)
Out[85]: (nan+nan*j)

while the correct answer is 1.0

Solution: Program tanh as:

tanh(z) = (1.0 - exp(-2.0*z)) / (1.0 + exp(-2.0*z))

### PR DESCRIPTION
**Pull request by charris**
Simplified and rebased version of #3010.

np.tanh(1000+0j) gives nan+nan_j instead of 1.0+0j.
np.tan(0+1000j) gives nan+nan_j instead of 1j.

I've imported the implementation for ctanh from FreeBSD's math library
which handles large arguments correctly and fixes this bug and the
equivalent bug in np.tan.

The problem here is that importing the function into npy_math_complex.c
and adding a config check causes us to use the implementation from glibc
on Linux, which also has this bug. Although it seems to have been fixed
in glibc last April
(http://sourceware.org/bugzilla/show_bug.cgi?id=11521).

I see several things that could be done here, the easiest is probably to
use our version of ctanh unconditionally. Although there are a multitude
of ways to go about that, for instance should I remove the
implementation from npy_math_complex.c and just place it directly in
umath/funcs.inc, where the buggy version was? Or should I just remove
the config check and add a note about it somewhere?

This PR also uses more of the system complex functions if they are available. 

Closes #2321.

### CODE REVIEW

**Comment by charris** [comment]
This PR needs some style cleanups, but I've put those into a separate PR to keep this easier to review. Except for some cleanups that were already made. Same with creating new files and moving functions around.

The `cproj` function is in complex.h and we don't have a matching ufunc. It would probably be good to add that function, and maybe `conj` also just so the library is complete.

As far as testing the numpy fallback functions, I'm inclined to add a macro that will force the numpy versions to be used, so they can be tested with the python tests. Suggestions for alternatives are welcome.

Small changes in function values are likely to result, so this should probably be mentioned in the release notes if accepted.

**Comment by charris** [comment]
Looks like configuration testing is needed. Bummer.

**Comment by charris** [comment]
Still playing with this.

File: numpy/core/include/numpy/npy_math.h [hunk 1]
```diff
@@ -258,8 +261,11 @@ float npy_copysignf(float x, float y);
 float npy_nextafterf(float x, float y);
 float npy_spacingf(float x);
 
+float npy_ldexpf(float x, int exp);
```
Comment (ewmoore) [inline numpy/core/include/numpy/npy_math.h]: These are already in master.

**Comment by ewmoore** [comment]
This looks good to me.  Thanks for taking the time to clean this up.  I allowed my pull request to get way out of hand, but I think parts of this are still good improvements.

File: numpy/core/include/numpy/npy_math.h (hunk 1 above)
Comment (charris) [inline numpy/core/include/numpy/npy_math.h]: Fixed, thanks.

**Pull request merged**